### PR TITLE
Create CachedResolvingDecoder and related classes and method

### PR DESCRIPTION
**Pull request by guawang**
Create CachedResolvingDecoder and related classes and method to help espresso-pub works on different version of Avro.
This PR is only for AVRO14, and will submit other PRs for avro1.5-1.10 if this PR passes.

### CODE REVIEW

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2020 LinkedIn Corp.
+ * Licensed under the BSD 2-Clause License (the "License").
+ * See License in the project root for license information.
+ */
+
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java:8]: this class was not copied from avro so i dont think the avro license header belongs here

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import com.linkedin.avroutil1.compatibility.avro14.parsing.CachedResolvingGrammarGenerator;
+import com.linkedin.avroutil1.compatibility.avro14.parsing.Symbol;
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.avro.Schema;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.Decoder;
+
+
+public class CachedResolvingDecoder extends ResolvingDecoder {
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java:35]: please explain the purpose of the class - something like "an extension of ResolvingDecoder that caches the grammar given  a pair of writer and reader schemas, as opposed to the parent class that re-generates the grammer on each call to DatumReader.read()". or something.

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  private static final ConcurrentHashMap<Schema, ConcurrentHashMap<Schema, Symbol>>
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java:36]: create a key class that holds a schema pair. Map<Schema, Map<Schema, Symbol>> --> Map< {Schema, Schema}, Symbol>

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+   * {@link #resolve(Schema, Schema)}.
+   * @param resolver  The resolver to use.
+   * @param in  The underlying decoder.
+   * @throws IOException
+   */
+  public CachedResolvingDecoder(Object resolver, Decoder in)
+      throws IOException {
+    super(resolver, in);
+  }
+  /*skips the symbol and returns the size of the string to copy*/
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java:55]: better documentation on what these methods are, and what the intended use is?

Comment (radai-rosenblatt) [reply helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java:55]: maybe call these skipString, skipBytes etc ?

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  /*returns the string data*/
+  public void readStringData(byte[] bytes, int start, int len) throws IOException {
+    in.readFixed(bytes, start, len);
+  }
+  /*returns the bytes data*/
+  public void readBytesData(byte[] bytes, int start, int len) throws IOException {
+    in.readFixed(bytes, start, len);
+  }
+
+  public boolean isBinaryDecoder() {
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java:74]: what would this method be used for? what does it matter if the decoder is consuming binary payload or json payload ?

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/ResolvingDecoder.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+import com.linkedin.avroutil1.compatibility.avro14.parsing.ResolvingGrammarGenerator;
+import com.linkedin.avroutil1.compatibility.avro14.parsing.Symbol;
+import java.io.IOException;
+import org.apache.avro.AvroTypeException;
+import org.apache.avro.Schema;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+
+/**
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/ResolvingDecoder.java:34]: i assume this was copied out of avro as-is ? if so - why is this in the same commit as above?

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/parsing/ResolvingGrammarGeneratorPr.java [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.Encoder;
+import org.codehaus.jackson.JsonNode;
+
+
+/**
+ * The class that generates a resolving grammar to resolve between two
+ * schemas.
+ */
+public class ResolvingGrammarGeneratorPr extends ValidatingGrammarGenerator {
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/parsing/ResolvingGrammarGeneratorPr.java:45]: why does this class exist? what does it do differently to its parent class?

File: helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java (hunk 5 above)
Comment (guawang) [reply helper/impls/helper-impl-14/src/main/java/com/linkedin/avroutil1/compatibility/avro14/codec/CachedResolvingDecoder.java]: In espresso internal EspressoDatumConverter, if both input and output are binary, decoders can optimize the copy

**Pull request closed without merging**